modules/imgproc/test/test_color.cpp

```diff
@@ -1935,20 +2090,342 @@ TEST(Imgproc_ColorLab_Full, accuracy)
+        go = CV_DESCALE(coeffs[3]*xx + coeffs[4]*yy + coeffs[5]*zz, shift);
+        bo = CV_DESCALE(coeffs[6]*xx + coeffs[7]*yy + coeffs[8]*zz, shift);
+
+        ro = max(0, min((int)INV_GAMMA_TAB_SIZE-1, ro));
+        go = max(0, min((int)INV_GAMMA_TAB_SIZE-1, go));
+        bo = max(0, min((int)INV_GAMMA_TAB_SIZE-1, bo));
+
+        ro = tab[ro];
+        go = tab[go];
+        bo = tab[bo];
+
+        dst_row[x*cn    ] = saturate_cast<uchar>(bo);
+        dst_row[x*cn + 1] = saturate_cast<uchar>(go);
+        dst_row[x*cn + 2] = saturate_cast<uchar>(ro);
+        if(cn == 4) dst_row[x*cn + 3] = 255;
+    }
+
+    return n;
+}
+
+int row8uLabChoose(const uchar* src_row, uchar *dst_row, int n, bool forward, int blue_idx, bool srgb)
+{
+    if(forward)
+        return row8uRGB2Lab(src_row, dst_row, n, 3, blue_idx, srgb);
+    else
+        return row8uLab2RGB(src_row, dst_row, n, 3, blue_idx, srgb);
+}
+
+
+TEST(Imgproc_ColorLab_Full, bitExactness)
+{
+    int codes[] = { CV_BGR2Lab, CV_RGB2Lab, CV_LBGR2Lab, CV_LRGB2Lab,
+                    CV_Lab2BGR, CV_Lab2RGB, CV_Lab2LBGR, CV_Lab2LRGB};
+    string names[] = { "CV_BGR2Lab", "CV_RGB2Lab", "CV_LBGR2Lab", "CV_LRGB2Lab",
+                       "CV_Lab2BGR", "CV_Lab2RGB", "CV_Lab2LBGR", "CV_Lab2LRGB" };
+
+    // need to be recalculated each time we change Lab algorithms, RNG or test system
+    const int nIterations = 8;
+    uint32_t hashes[] = {
+        0xca7d94c4, 0x34aeb79a, 0x7272c2cf, 0x62c2efed, 0x047cab77, 0x5e8dfb85, 0x10fed613, 0x34d2f4aa,
+        0x048bea9a, 0xbbe20ef2, 0x3274e88f, 0x710e9272, 0x9fd6cd59, 0x69d67639, 0x04742095, 0x9ef2b60b,
+        0x75b78f5b, 0x3fda9801, 0x374cc472, 0x3239e8ad, 0x94749b2d, 0x9362ac0c, 0xa4d7dd36, 0xe25ef694,
+        0x51d1b01d, 0xb0f6e3f5, 0x2b72a228, 0xb7429fa0, 0x799ba6bd, 0x2141d3d2, 0xb4dde471, 0x813b6e0f,
+        0x9c029161, 0xb51eb5ec, 0x460c3a09, 0x27724f63, 0xb446c9a8, 0x3adf1b61, 0xe6b0d30f, 0xd1078779,
+        0xfaa7525b, 0x5b6ea158, 0xdf3511f7, 0xf01dc02d, 0x5c663841, 0xce611ed4, 0x758ad851, 0xa43c3a1c,
+        0xed30f68c, 0xcb6babd9, 0xf38262b5, 0x608cb3db, 0x13425e5a, 0x6dc5fdc7, 0x9519090a, 0x87aa73d0,
+        0x8e9bf980, 0x46b98728, 0x0064591c, 0x7e1efc9b, 0xf0ec2465, 0x89a75c8d, 0x0d162fa7, 0xffea7a2f,
+    };
+
+    RNG rng(0);
+    // blueIdx x srgb x direction
+    bool next = true;
+    for(int c = 0; next && c < 8; c++)
+    {
+        int v = c;
+        int  blueIdx = (v % 2 != 0) ? 2 : 0; v /=2;
+        bool    srgb = (v % 2 == 0); v /= 2;
+        bool forward = (v % 2 == 0);
+
+        for(int iter = 0; next && iter < nIterations; iter++)
+        {
+            Mat probe(256, 256, CV_8UC3), result;
+            rng.fill(probe, RNG::UNIFORM, 0, 255, true);
+
+            cvtColor(probe, result, codes[c]);
+
+            uint32_t h = adler32(result);
+
+            if(h != hashes[c*nIterations + iter])
+            {
+                initLabTabs();
```

Comment:
We should dump information about mode, iteration, actual hash - at least to make updates of hash table above.